server/src-lib/Hasura/Db.hs

```diff
@@ -36,14 +36,16 @@ data PGExecCtx
   }
 
 class (MonadError QErr m) => MonadTx m where
+  liftTxWith :: Q.TxE QErr a -> Maybe Q.TxAccess -> m a
```

Comment:
I think this probably isn’t the right thing to do here since `MonadTx` describes a class of computations that are already in a transaction, so `liftTxWith` can’t do anything meaningful if the transaction has already begun. However, I think Postgres actually allows changing a transaction’s access mode in the middle of a transaction via [`SET TRANSACTION READ ONLY`](https://www.postgresql.org/docs/12/sql-set-transaction.html). Therefore, `run_sql` can probably just execute that directly to switch into read-only mode, then switch back to read/write mode afterwards.